torchtune/_cli/run.py

```diff
@@ -0,0 +1,159 @@
+
+    def __init__(self, subparsers):
+        super().__init__()
+        self._parser = subparsers.add_parser(
+            "run",
+            prog="tune run",
+            help="Run a recipe. For distributed recipes, this supports all torchrun arguments.",
+            description="Run a recipe. For distributed recipes, this supports all torchrun arguments.",
+            usage="tune run [TORCHRUN-OPTIONS] <recipe> --config <config> [RECIPE-OPTIONS]",
+            epilog=textwrap.dedent(
+                """\
+                examples:
+
+                    # Run a finetuning recipe on a single device w/ default values
+                    $ tune run lora_finetune_single_device --config llama2/7B_lora_single_device
+
+                    # Run a finetuning recipe in a distributed fashion using torchrun w/ default values
+                    $ tune run --nproc_per_node 4 full_finetune_distributed --config llama2/7B_full_finetune_distributed
+
+                    # Override a parameter in the config file and specify a number of GPUs for torchrun
+                    $ tune run lora_finetune_single_device \
+                        --config llama2/7B_lora_single_device \
+                        model.lora_rank=16 \
+
+                Remember, you can use `tune cp` to copy a default recipe/config to your local dir and modify the values.
+                """
+            ),
+            formatter_class=argparse.RawTextHelpFormatter,
+        )
+        self._add_arguments()
+        self._parser.set_defaults(func=self._run_cmd)
+
+    def _add_arguments(self) -> None:
+        """Add arguments to the parser."""
+        torchrun_argparser = get_torchrun_args_parser()
+        # Add torchrun arguments to the parser
+        for action in torchrun_argparser._actions:
+            if action.dest == "training_script":
+                action.dest = "recipe"
+                action.help = """
+                    Name or path to recipe to be launched followed by args.
+                    For a list of all possible recipes, run `tune ls`."""
+            elif action.dest == "training_script_args":
+                action.dest = "recipe_args"
+                action.help = "Args to be passed to the recipe."
+            elif action.dest == "help":
+                continue
+            self._parser._add_action(action)
+
+    def _run_distributed(self, args: argparse.Namespace):
+        """Run a recipe with torchrun."""
+        # TODO (rohan-varma): Add check that nproc_per_node <= cuda device count. Currently,
+        # we don't do this since we test on CPUs for distributed. Will update once multi GPU CI is supported.
+        print("Running with torchrun...")
+        # Have to reset the argv so that the recipe can be run with the correct arguments
+        args.__dict__["training_script"] = args.__dict__.pop("recipe")
+        args.__dict__["training_script_args"] = args.__dict__.pop("recipe_args")
```

Comment:
Do we actually need to use `__dict__` on the LHS of these two lines?

Reply:
Nah, fixed.